### PR TITLE
Fix index joins using inequality predicates

### LINKED ISSUE #1105: Join predicates which use inequalities are incorrect when using index joins
Until Terrier adds functionality for specifying the join algorithm, reproducing this bug will be a bit contrived. But you can do the following,

```
CREATE TABLE xxx (id INT PRIMARY KEY, val INT);
CREATE TABLE yyy (id INT PRIMARY KEY, val INT);
INSERT INTO xxx VALUES  (3,3), (4,4), (5,5);
INSERT INTO yyy VALUES (3, 3), (4,4);
SELECT * FROM xxx INNER JOIN yyy ON xxx.val > yyy.val;
```

which produces the following correct output using a nested loop join: 
```
 id | val | id | val 
----+-----+----+-----
  4 |   4 |  3 |   3
  5 |   5 |  3 |   3
  5 |   5 |  4 |   4
(3 rows)
```
However, the following query,

`SELECT * FROM xxx INNER JOIN yyy ON xxx.id > yyy.id;` 

which should produce the same output, instead produces:
```
 id | val | id | val 
----+-----+----+-----
(0 rows)

```
because it uses an index join.

### PR DESCRIPTION
**Pull request by wgy8283335**
Fix #1105.

### CODE REVIEW

**Comment by wgy8283335** [comment]
I've done some tests, it seems to work well.
```
terrier=# create database test;
CREATE DATABASE
terrier=# 
terrier=# \c test;
psql (10.12 (Ubuntu 10.12-0ubuntu0.18.04.1), server 9.5devel)
You are now connected to database "test" as user "terrier".
test=# 
test=# CREATE TABLE xxx (idx INT PRIMARY KEY, valx INT);
CREATE TABLE
test=# CREATE TABLE yyy (idy INT PRIMARY KEY, valy INT);
CREATE TABLE
test=# INSERT INTO xxx VALUES (3,3), (4,4), (5,5);
INSERT 0 3
test=# INSERT INTO yyy VALUES (3, 3), (4,4);
INSERT 0 2
test=# SELECT * FROM xxx INNER JOIN yyy ON xxx.idx > yyy.idy;
 idy | valy | idx | valx 
-----+------+-----+------
   3 |    3 |   4 |    4
   3 |    3 |   5 |    5
   4 |    4 |   5 |    5
(3 rows)

test=# SELECT * FROM xxx INNER JOIN yyy ON xxx.idx < yyy.idy;
 idy | valy | idx | valx 
-----+------+-----+------
   4 |    4 |   3 |    3
(1 row)

test=# SELECT * FROM xxx INNER JOIN yyy ON 5 > yyy.idy;
 idy | valy | idx | valx 
-----+------+-----+------
   3 |    3 |   3 |    3
   4 |    4 |   3 |    3
   3 |    3 |   4 |    4
   4 |    4 |   4 |    4
   3 |    3 |   5 |    5
   4 |    4 |   5 |    5
(6 rows)

test=# SELECT * FROM xxx INNER JOIN yyy ON 2 < yyy.idy;
 idy | valy | idx | valx 
-----+------+-----+------
   3 |    3 |   3 |    3
   4 |    4 |   3 |    3
   3 |    3 |   4 |    4
   4 |    4 |   4 |    4
   3 |    3 |   5 |    5
   4 |    4 |   5 |    5
(6 rows)

test=# SELECT * FROM xxx INNER JOIN yyy ON xxx.idx < 6;
 idy | valy | idx | valx 
-----+------+-----+------
   3 |    3 |   3 |    3
   3 |    3 |   4 |    4
   3 |    3 |   5 |    5
   4 |    4 |   3 |    3
   4 |    4 |   4 |    4
   4 |    4 |   5 |    5
(6 rows)

test=# SELECT * FROM xxx INNER JOIN yyy ON xxx.idx > 3;
 idy | valy | idx | valx 
-----+------+-----+------
   3 |    3 |   4 |    4
   3 |    3 |   5 |    5
   4 |    4 |   4 |    4
   4 |    4 |   5 |    5
(4 rows)
```

**Comment by wgy8283335** [comment]
Always fail in the microbenchmark.
image
image

**Comment by wgy8283335** [comment]
image
@mbutrovich , all of the checks have passed.

**Comment by wgy8283335** [comment]
@apavlo Dear Processor Pavlo, The CI had succeeded two days ago, but I find the state of ci changes from succeeding to failure today. I'm a little curious about that. Does the CI repeatedly execute even after it succeed?

**Comment by mbutrovich** [comment]
CI will re-run any time something is merged into master, yes.

Also are you able to add some trace-file tests of the examples you posted and then we can get this reviewed. Thank you for the effort so far!

**Comment by wgy8283335** [comment]
> CI will re-run any time something is merged into master, yes.
> 
> Also are you able to add some trace-file tests of the examples you posted and then we can get this reviewed. Thank you for the effort so far!

Sure. Do you mean I should add some tests in the "select.test" file, as below?
image

**Comment by mbutrovich** [comment]
Yep I think that's the right place for them. Thanks!

**Comment by wgy8283335** [comment]
> Yep I think that's the right place for them. Thanks!

It's my pleasure. I'm writing the tests now, and I encounter a problem: how could I calculate the hash value? Which utility could generate the hash value?
image

**Comment by mbutrovich** [comment]
Thanks for adding those tests! I'll try to find you an appropriate reviewer.

**Comment by wgy8283335** [comment]
> Thanks for adding those tests! I'll try to find you an appropriate reviewer.
Hello, my friend. I'm just a little concerned about the reviewer. Have you found someone to review the code?

File: script/testing/junit/traces/select.test [hunk 1]
```diff
@@ -580,6 +580,55 @@ SELECT str_a_val NOT LIKE 'AbCdEf' AS result FROM like_tbl WHERE str_a_val IS NU
...
+CREATE TABLE yyy (idy INT PRIMARY KEY, valy INT);
+
+statement ok
+INSERT INTO xxx VALUES (3,3), (4,4), (5,5);
+
+statement ok
+INSERT INTO yyy VALUES (3, 3), (4,4);
+
+query I nosort
+SELECT * FROM xxx INNER JOIN yyy ON xxx.idx > yyy.idy;
```
Comment (17zhangw) [inline script/testing/junit/traces/select.test:597]: Thank you for the general tests! These generally look pretty good. For verifiability, can you also add a test that flips the two tables such that `xxx.idx` appears on the right side of the comparison as opposed to the left?

File: src/optimizer/index_util.cpp [hunk 2]
```diff
@@ -88,6 +88,7 @@ bool IndexUtil::CheckPredicates(
   // To concatenate/shrink ranges, we would need to be able to compare TransientValues.
   std::unordered_map<catalog::indexkeycol_oid_t, planner::IndexExpression> open_highs;  // <index, low start>
   std::unordered_map<catalog::indexkeycol_oid_t, planner::IndexExpression> open_lows;   // <index, high end>
+  int flag = 1;
```
Comment (17zhangw) [inline src/optimizer/index_util.cpp:91]: Can this variable be named something more descriptive?

**Review by 17zhangw — CHANGES_REQUESTED** [review]
Generally looks good! Just some small comments. Thanks!

File: script/testing/junit/traces/select.test (hunk 1 above)
Comment (wgy8283335) [reply script/testing/junit/traces/select.test:597]: > Thank you for the general tests! These generally look pretty good. For verifiability, can you also add a test that flips the two tables such that `xxx.idx` appears on the right side of the comparison as opposed to the left?

I‘ve added one test.

File: src/optimizer/index_util.cpp (hunk 2 above)
Comment (wgy8283335) [reply src/optimizer/index_util.cpp]: > Can this variable be named something more descriptive?

Yes, I have changed the name from flag to left_side and change the type.

**Review by 17zhangw — APPROVED** [review]
LGTM!

**Pull request merged**